src/algebra/is_prime_pow.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2022 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import algebra.associated
+import data.nat.factorization
+
+/-!
+# Prime powers
+
+This file deals with prime powers: numbers which are positive integer powers of a single prime.
+-/
+
+variables {R : Type*} [comm_monoid_with_zero R] (n p : R) (k : ℕ)
+
+/-- `n` is a prime power if there is a prime `p` and a positive natural `k` such that `n` can be
+written as `p^k`. -/
+def is_prime_pow : Prop :=
+  ∃ (p : R) (k : ℕ), prime p ∧ 0 < k ∧ p ^ k = n
+
+lemma is_prime_pow_def :
+  is_prime_pow n ↔ ∃ (p : R) (k : ℕ), prime p ∧ 0 < k ∧ p ^ k = n := iff.rfl
+
+/-- An equivalent definition for prime powers: `n` is a prime power iff there is a prime `p` and a
+natural `k` such that `n` can be written as `p^(k+1)`. -/
+lemma is_prime_pow_iff_pow_succ :
+  is_prime_pow n ↔ ∃ (p : R) (k : ℕ), prime p ∧ p ^ (k + 1) = n :=
+(is_prime_pow_def _).trans
+⟨λ ⟨p, k, hp, hk, hn⟩, ⟨_, _, hp, by rwa [nat.sub_add_cancel hk]⟩,
+  λ ⟨p, k, hp, hn⟩, ⟨_, _, hp, nat.succ_pos', hn⟩⟩
+
+lemma not_is_prime_pow_zero [no_zero_divisors R] :
+  ¬ is_prime_pow (0 : R) :=
+begin
+  simp only [is_prime_pow_def, not_exists, not_and', and_imp],
+  intros x n hn hx,
+  rw pow_eq_zero hx,
+  simp,
+end
+
+lemma not_is_prime_pow_one : ¬ is_prime_pow (1 : R) :=
+begin
+  simp only [is_prime_pow_def, not_exists, not_and', and_imp],
+  intros x n hn hx ht,
+  exact ht.not_unit (is_unit_of_pow_eq_one x n hx hn),
+end
+
+lemma prime.is_prime_pow {p : R} (hp : prime p) : is_prime_pow p :=
+⟨p, 1, hp, zero_lt_one, by simp⟩
+
+lemma is_prime_pow.pow {n : R} (hn : is_prime_pow n)
+  {k : ℕ} (hk : k ≠ 0) : is_prime_pow (n ^ k) :=
+let ⟨p, k', hp, hk', hn⟩ := hn in ⟨p, k * k', hp, mul_pos hk.bot_lt hk', by rw [pow_mul', hn]⟩
+
+lemma unique_is_prime_pow {R : Type*} [cancel_comm_monoid_with_zero R] {p₁ p₂ : R} {k₁ k₂ : ℕ}
```

Comment:
this is `associated.of_pow_associated_of_prime`

Reply:
Yup you're right, I moved this and cleaned it and forgot to remove the old version!